File: 29Items-Projects/7Algorithms/6-Backtracking/python/tug_of_war.py
"""Tug of War — partition set into two subsets with minimum absolute difference."""


class TugOfWar:
    def __init__(self, arr: list[int]):
        if len(arr) < 2:
            raise ValueError("Array must have at least 2 elements")
        self.arr = arr
        self.n = len(arr)
        self.min_diff = float("inf")
        self.best_set1: list[int] = []
        self.best_set2: list[int] = []

    def solve_and_print(self) -> None:
        print(f"Input: {self.arr}")
        self._backtrack([], [], 0)
        print(f"Minimum difference: {self.min_diff}\n")
        print(f"Set 1: {self.best_set1} (sum = {sum(self.best_set1)})")
        print(f"Set 2: {self.best_set2} (sum = {sum(self.best_set2)})")

    def _backtrack(self, set1: list[int], set2: list[int], idx: int) -> None:
        if idx == self.n:
            diff = abs(sum(set1) - sum(set2))
            if diff < self.min_diff:
                self.min_diff = diff
                self.best_set1 = set1[:]
                self.best_set2 = set2[:]
            return

        if abs(sum(set1) - sum(set2)) > self.min_diff:
            return

        set1.append(self.arr[idx])
        self._backtrack(set1, set2, idx + 1)
        set1.pop()

        set2.append(self.arr[idx])
        self._backtrack(set1, set2, idx + 1)
        set2.pop()

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/tugofwar/TugOfWar.java
package tugofwar;

import java.util.*;

public class TugOfWar {

    private final int[] arr;
    private final int n;
    private int minDiff = Integer.MAX_VALUE;
    private List<Integer> bestSet1;
    private List<Integer> bestSet2;

    public TugOfWar(int[] arr) {
        if (arr == null || arr.length < 2)
            throw new IllegalArgumentException("Array must have at least 2 elements");
        this.arr = arr.clone();
        this.n = arr.length;
    }

    public void solveAndPrint() {
        System.out.println("Input: " + Arrays.toString(arr));

        backtrack(new ArrayList<>(), new ArrayList<>(), 0);

        System.out.printf("Minimum difference: %d%n%n", minDiff);
        int sum1 = bestSet1.stream().mapToInt(x -> x).sum();
        int sum2 = bestSet2.stream().mapToInt(x -> x).sum();
        System.out.printf("Set 1: %s (sum = %d)%n", bestSet1, sum1);
        System.out.printf("Set 2: %s (sum = %d)%n", bestSet2, sum2);
    }

    private void backtrack(List<Integer> set1, List<Integer> set2, int index) {
        if (index == n) {
            int diff = Math.abs(
                    set1.stream().mapToInt(x -> x).sum() -
                    set2.stream().mapToInt(x -> x).sum());
            if (diff < minDiff) {
                minDiff = diff;
                bestSet1 = new ArrayList<>(set1);
                bestSet2 = new ArrayList<>(set2);
            }
            return;
        }

        // Prune: if current difference already exceeds best, skip
        if (Math.abs(set1.stream().mapToInt(x -> x).sum() -
                     set2.stream().mapToInt(x -> x).sum()) > minDiff) {
            return;
        }

        set1.add(arr[index]);
        backtrack(set1, set2, index + 1);
        set1.remove(set1.size() - 1);

        set2.add(arr[index]);
        backtrack(set1, set2, index + 1);
        set2.remove(set2.size() - 1);
    }
}
